lib/rust/ensogl/core/src/animation/loops.rs

```diff
@@ -251,26 +285,35 @@ impl InitializedTimeInfo {
 
 /// Callback for an animation frame.
 pub type OnFrameClosure = impl FnMut(Duration);
-fn on_frame_closure(frp: &Frp, callbacks: &callback::registry::Copy1<Duration>) -> OnFrameClosure {
-    let callbacks = callbacks.clone_ref();
+fn on_frame_closure(
+    frp: &Frp,
+    before_animations: &callback::registry::Copy1<TimeInfo>,
+    animations: &callback::registry::Copy1<FixedFrameRateStep<TimeInfo>>,
+) -> OnFrameClosure {
+    let before_animations = before_animations.clone_ref();
+    let animations = animations.clone_ref();
     let output = frp.private.output.clone_ref();
-
     let mut time_info = InitializedTimeInfo::default();
     let h_cell = Rc::new(Cell::new(callback::Handle::default()));
+    let fixed_fps_sampler = Rc::new(RefCell::new(FixedFrameRateSampler::default()));
+
     move |frame_time: Duration| {
         let time_info = time_info.next_frame(frame_time);
         let on_frame_start = output.on_frame_start.clone_ref();
         let on_before_animations = output.on_before_animations.clone_ref();
         let on_after_animations = output.on_after_animations.clone_ref();
         let frame_end = output.frame_end.clone_ref();
         let output = output.clone_ref();
-        let callbacks = callbacks.clone_ref();
+        let before_animations = before_animations.clone_ref();
+        let animations = animations.clone_ref();
         let _profiler = profiler::start_debug!(profiler::APP_LIFETIME, "@on_frame");
+        let fixed_fps_sampler = fixed_fps_sampler.clone_ref();
 
         TickPhases::new(&h_cell)
             .then(move || on_frame_start.emit(frame_time))
             .then(move || on_before_animations.emit(time_info))
-            .then(move || callbacks.run_all(frame_time))
+            .then(move || before_animations.run_all(time_info))
+            .then(move || fixed_fps_sampler.borrow_mut().run(time_info, |t| animations.run_all(t)))
```

Comment:
Using `FixedFrameRateSampler` for all animations (including easings) is IMO not a good idea. It will introduce visible stutter whenever there is a slight frame refresh delay, even on 60Hz displays. Especially with freesync being commonplace. Also, many current displays are way above 60Hz (even m1 macbook uses 120Hz screen).

Here is a good example of what it might look like
https://www.testufo.com/freesync#photo=quebec.jpg&pps=960&framerate=struggle&compare=1

Fixed timestep should only be used for animations that use differential equation solver (usually physics). Even then, ideally those should be always simulated 1 frame in advance, then two frames needs to be interpolated based on real display frame time. Otherwise it will be visibly stuttering.


I see that fixed frame here is used mainly to be able to calculate previous frame's time on first animation frame. I don't think it's necessary though. A normal frame stepper should also work fine there, as long as we initialize animation start time properly - it should always be set to the current frame time (NOT `performance.now()`) at the time the callback is added. To do that, we need to promote `time_info` variable into a member of `LoopRegistry`, so it can be used there. That way the first time the animation is actually processed, it will already have `dt > 0`. This would also save us a branch on setting the initialization time, because it is guaranteed to be initialized when the callback actually fires.

Reply:
Comment from our discussion - the struct name is confusing. This is not really fixed frame rate sampler and it works correctly now. I will fix the struct naming + add more docs in the next PR.